File: golang/reshape_the_matrix.go
package main

// Space : O(c)
// Time  : O(r*c)

func matrixReshape(mat [][]int, r int, c int) [][]int {
    row := len(mat)
    col := len(mat[0])
    
    if row*col != r*c {
        return mat
    }
    
    ans, temp := [][]int{}, []int{}
    for y := 0; y < row; y++ {
        for x := 0; x < col; x++ {
            temp = append(temp, mat[y][x])
            if len(temp) == c {
                ans = append(ans, temp)
                temp = []int{}
            }
        }
    }
    
    return ans
}

File: python3/reshape_the_matrix.py
# Space : O(1)
# Time  : O(r*c)

class Solution:
    def matrixReshape(self, mat: List[List[int]], r: int, c: int) -> List[List[int]]:
        row = len(mat)
        col = len(mat[0])

        if row*col != r*c:
            return mat

        ans = [[] for _ in range(r)]
        point = 0
        for y in range(row):
            for x in range(col):
                ans[point].append(mat[y][x])
                if len(ans[point]) == c:
                    point += 1

        return ans
